### PR TITLE
Feature/mc dropout (WIP)

### PR DESCRIPTION
**Pull request by tom2208**
Working on MC Dropout here.
Ye et al. are using ResNet50. So I will approach this model first.
Thier code: https://github.com/shijing001/Uncertainty_calibration_audio_classifiers/blob/main/models/resnet.py

Another approach:
Write own resnet class that derives from the trochvision implementation and override the forward function: https://github.com/pytorch/vision/blob/61763fa955ef74077a1d3e1aa5da36f7c606943a/torchvision/models/resnet.py#L197-L213

Another approach:
Use of forward hooks: https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html
But i am not sure if this could replicate the experiments of Ye et al.

Another approach:
timm libary provides ResNet with Dropout implemented: https://github.com/huggingface/pytorch-image-models?tab=readme-ov-file#models
But its not very well documented.

Roadmap:

- [x] researched methods of modifying a ResNet to implement dropout
- [x] get the new implementation method to work
    - [x] Try to modify the pretrained and trained ResNet (-> not a successful approach)
    - [x] try approach: overriding the forward method in a subclass
    - [x] make it work without pretrained
    - [x] make it work with pretrained 
- [x] implement the dropouts like Ye et al.
- [ ] try to replicate the same experiments like Ye et al. with ResNet on HSN
- [ ] try to improve on that

### CODE REVIEW

**Comment by raphaelschwinger** [comment]
@tom2208 This PR is still WIP, right? Could you prepend the title with "WIP"?
Also, it would be great to add a roadmap where you tick off things you already did and what things are still to do ;)

**Comment by tom2208** [comment]
I just pursued the second approach which is overriding necessary functions of the original resnet class. This is also what Ye et al. did. I just added the dropout like Ye et al. for now. I am not sure if this is the ideal way. But before trying something else I want to make sure the current code runs just fine.
